File: src/melody_sync/core/scanner/discovery.py
from pathlib import Path

SUPPORTED_AUDIO_SUFFIXES: set[str] = {
    ".aac",
    ".flac",
    ".m4a",
    ".mp3",
    ".ogg",
    ".opus",
    ".wav",
}


def discover(directory: Path) -> list[Path]:
    """
    Encontra todos os arquivos de áudio suportados em um diretório.

    Args:
        directory:
            Diretório da biblioteca musical.

    Returns:
        Lista de músicas encontradas.
    """

    if not directory.exists():
        raise FileNotFoundError(directory)

    if not directory.is_dir():
        raise NotADirectoryError(directory)

    return sorted(
        file
        for file in directory.rglob("*")
        if file.is_file()
        and file.suffix.lower() in SUPPORTED_AUDIO_SUFFIXES
    )

File: melody-sync-core/src/main/kotlin/com/melodysync/scanner/Discovery.kt
package com.melodysync.scanner

import java.nio.file.Files
import java.nio.file.NoSuchFileException
import java.nio.file.NotDirectoryException
import java.nio.file.Path

val SUPPORTED_AUDIO_SUFFIXES: Set<String> = setOf(
    ".aac",
    ".flac",
    ".m4a",
    ".mp3",
    ".ogg",
    ".opus",
    ".wav",
)

fun discover(directory: Path): List<Path> {
    if (!Files.exists(directory)) throw NoSuchFileException(directory.toString())
    if (!Files.isDirectory(directory)) throw NotDirectoryException(directory.toString())

    return Files.walk(directory).use { paths ->
        paths.filter { file ->
            Files.isRegularFile(file) && file.fileName.toString().lowercase().substringAfterLast('.').let { ".$it" in SUPPORTED_AUDIO_SUFFIXES }
        }.sorted().toList()
    }
}
